application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/LicenseRenewListener.java

```diff
@@ -0,0 +1,208 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.extension.ExtensionId;
+import org.xwiki.extension.InstalledExtension;
+import org.xwiki.extension.event.ExtensionEvent;
+import org.xwiki.extension.event.ExtensionInstalledEvent;
+import org.xwiki.extension.event.ExtensionUpgradedEvent;
+import org.xwiki.extension.repository.InstalledExtensionRepository;
+import org.xwiki.observation.EventListener;
+import org.xwiki.observation.event.Event;
+
+import com.xwiki.licensing.License;
+import com.xwiki.licensing.LicenseUpdater;
+import com.xwiki.licensing.LicensedExtensionManager;
+import com.xwiki.licensing.LicensedFeatureId;
+import com.xwiki.licensing.Licensor;
+
+/**
+ * Listen to licensed extensions upgrades or install in order to trigger a license renewal if needed. A license renew
+ * it's needed if licensed dependencies (direct or transitive) have changed between versions, or if the currnt license
+ * has outdated feature ids. Listen also for the install event in case a license was already generated for an older
+ * uninstalled version.
+ *
+ * @version $Id$
+ * @since 1.27
+ */
+@Component
+@Named(LicenseRenewListener.NAME)
+@Singleton
+public class LicenseRenewListener implements EventListener
+{
+    protected static final String NAME = "com.xwiki.licensing.internal.LicenseRenewListener";
+
+    protected static final List<Event> EVENTS =
+        Arrays.asList(new ExtensionInstalledEvent(), new ExtensionUpgradedEvent());
+
+    @Inject
+    private LicensedExtensionManager licensedExtensionManager;
+
+    @Inject
+    private InstalledExtensionRepository installedExtensionRepository;
+
+    @Inject
+    private Provider<Licensor> licensorProvider;
+
+    @Inject
+    private LicenseUpdater licenseUpdater;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public List<Event> getEvents()
+    {
+        return EVENTS;
+    }
+
+    @Override
+    public String getName()
+    {
+        return NAME;
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        // Retrieve license updates from store.
+        licenseUpdater.getLicensesUpdates();
+
+        ExtensionEvent extensionEvent = (ExtensionEvent) event;
+        // Only treat top level licensed extensions. If it's not a top level extension, then its license it's handled
+        // by another extension's license.
+        if (licensedExtensionManager.getLicensedExtensions().contains(extensionEvent.getExtensionId())
+            && isMandatoryLicensedExtension(extensionEvent.getExtensionId()))
+        {
+            InstalledExtension installedExtension = (InstalledExtension) source;
+            if (installedExtension == null) {
```

Comment:
Would be good to add a comment explaining when the install / upgrade event is triggered without passing the installed / upgraded extension. Looks strange. Also, this check is way faster than the previous one, so I'd check this first.